api/envoy/config/route/v3/route_components.proto

```diff
@@ -1712,6 +1712,27 @@ message RateLimit {
       Source source = 4 [(validate.rules).enum = {defined_only: true}];
     }
 
+    // The following descriptor entry is appended with a computed value.
+    // See :ref:`attributes <arch_overview_attributes>` for the set of
+    // available attributes.
+    //
+    // .. code-block:: cpp
+    //
+    //   ("<descriptor_key>", "<expression_value>")
+    message Expression {
+      // The key to use in the descriptor entry.
+      string descriptor_key = 1 [(validate.rules).string = {min_len: 1}];
+
+      oneof expr_specifier {
```

Comment:
why is this a oneof?

Reply:
It's anticipating parsed_expr here which would introduce a dependency on expr proto. I can drop it since it's not needed right now.

Reply:
If you think we're going to need the `oneof` later, let's keep it for now.  We can't add it later without breaking the go proto API.

Reply:
Also, I believe that we are actually already using the CEL parsed expression proto in other places in the API, so this is probably not a new external dependency.  You might be able to just add it now.

Reply:
The dependency is coming from a filter. This is a core proto. Thoughts @htuch?

Reply:
If we don't want to depend on the CEL protos for the xDS core protos, another option might be to make this more extensible by simply adding a `TypedExpressionConfig`.  In the long term, we could maybe even migrate the existing options here to be extensions as well.

@envoyproxy/api-shepherds, thoughts?  I wonder if we should add something to the API review checklist in #14399 that prompts people to consider using `TypedExtensionConfig` instead of a `oneof`?

Reply:
From an Envoy binary perspective, are we adding this to core router, or does it only apply to rate limit filter? Hard to tell from config. If core, @kyessenov would be nice to have clarity on when CEL can a reasonable number of the checkmarks in https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md#new-external-dependencies. Whatever is reasonable, but if this is going to become a core data plane dependency, it's probably good to use this opportunity to check in. CC @envoyproxy/dependency-shepherds 

From an API perspective, I don't feel super strongly, CEL is a reasonably standard thing. One area to have some concern around is that using CEL is similar to regex; we're introducing dynamic evaluation of some expression, which ultimately might have security and inter-request isolation concerns. I don't know if this is in any way actionable, but certainly when communicating in comments/docs its worth pointing out what the runtime implications are.

Reply:
I believe the runtime change is scoped to global and local rate limit filters. So this is purely a question of a proto dependency, to avoid parsing overhead with a parsed expression-as-proto. Good point about dynamic evaluation, I'll document that. 

Expression evaluation is linear in both input and expression size. The input here is request metadata.

Reply:
The fact that rate limit stuff is in the main router is a historical artifact of before we had per-filter route config, etc. None of this should be here and I think there is a tracking issue to move it all out, but I don't think that should block this PR.

Reply:
OK, I think this forces moving expr package into common from extensions, since it gets pulled in the implementation.